### PR TITLE
Version conditional: make_valid

### PR DESCRIPTION
**Pull request by caspervdw**
Following up on #116 

MakeValid is available since GEOS 3.8.0

### CODE REVIEW

File: pygeos/geos.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+from functools import wraps
+
+
+class UnsupportedGEOSOperation(ImportError):
+    pass
+
+
+class requires_geos:
+    def __init__(self, version):
+        if version.count(".") == 1:
```
Comment (brendan-ward) [inline pygeos/geos.py]: @caspervdw since this annotation is used by pygeos contributors (rather than users), it seems reasonable to insist on standard usage (always <major>.<minor>.<patch>) instead of inferring patch component, e.g., 
```
if version.count(".") != 2:
    raise ValueError("version must be <major>.<minor>.<patch> format")
```

otherwise as written, this doesn't handle missing minor component...

Also seems like it would be good to parse version into a tuple here for internal use, so that `TypeError`s are caught in the annotation constructor (e.g., if can't be coerced to int) rather than in `__call__`

File: pygeos/constructive.py [hunk 2]
```diff
@@ -285,6 +287,22 @@ def extract_unique_points(geometry, **kwargs):
     return lib.extract_unique_points(geometry, **kwargs)
 
 
+@requires_geos("3.8.0")
+def make_valid(geometry, **kwargs):
+    """Repairs invalid geometries.
+
```
Comment (brendan-ward) [inline pygeos/constructive.py]: Would be good to add "Requires GEOS 3.8.0" to the docstring.

File: pygeos/test/test_constructive.py [hunk 3]
```diff
@@ -75,3 +75,25 @@ def test_snap_none():
...
     actual = pygeos.snap(geometry, point, tolerance=np.nan)
     assert actual is None
+
+
+@pytest.mark.skipif(pygeos.geos_version < (3, 8, 0), reason="GEOS < 3.8")
+def test_make_valid_none():
+    actual = pygeos.make_valid(None)
+    assert actual is None
+
+
```
Comment (brendan-ward) [inline pygeos/test/test_constructive.py]: Would be good to add test case for empty geometry (0d and 1d cases):

```
@pytest.mark.skipif(pygeos.geos_version < (3, 8, 0), reason="GEOS < 3.8")
@pytest.mark.parametrize(
    "geom,expected", [(empty, empty), ([empty], [empty])],
)
def test_make_valid_empty(geom, expected):
    assert pygeos.make_valid(geom) == expected
```

File: pygeos/test/test_constructive.py [hunk 4]
```diff
@@ -75,3 +75,25 @@ def test_snap_none():
...
+@pytest.mark.skipif(pygeos.geos_version < (3, 8, 0), reason="GEOS < 3.8")
+@pytest.mark.parametrize(
+    "geom,expected",
+    [
+        (pygeos.points(0, 0), pygeos.points(0, 0)),
+        (
+            Geometry("POLYGON((0 0, 1 1, 1 2, 1 1, 0 0))"),
+            Geometry("MULTILINESTRING ((0 0, 1 1), (1 1, 1 2))"),
```
Comment (brendan-ward) [inline pygeos/test/test_constructive.py]: A short comment noting what is not valid about the polygon would be helpful.

File: pygeos/test/test_constructive.py [hunk 5]
```diff
@@ -75,3 +75,25 @@ def test_snap_none():
...
+        ),
+    ],
```
Comment (brendan-ward) [inline pygeos/test/test_constructive.py]: A couple more test cases:

```
        # polygon with self-intersection (bowtie) converted to individual polygons
        (
            Geometry("POLYGON((0 0, 2 2, 2 0, 0 2, 0 0))"),
            Geometry("MULTIPOLYGON (((1 1, 0 0, 0 2, 1 1)), ((1 1, 2 2, 2 0, 1 1)))"),
        ),
        # first polygon is valid, second polygon has self-intersection
        (
            [
                Geometry("POLYGON((0 0, 2 2, 0 2, 0 0))"),
                Geometry("POLYGON((0 0, 2 2, 2 0, 0 2, 0 0))"),
            ],
            [
                Geometry("POLYGON((0 0, 2 2, 0 2, 0 0))"),
                Geometry(
                    "MULTIPOLYGON (((1 1, 0 0, 0 2, 1 1)), ((1 1, 2 2, 2 0, 1 1)))"
                ),
            ],
        ),

```

File: pygeos/test/test_constructive.py [hunk 6]
```diff
@@ -75,3 +75,25 @@ def test_snap_none():
...
+)
+def test_make_valid(geom, expected):
+    actual = pygeos.make_valid(geom)
+    assert actual == expected
```
Comment (brendan-ward) [inline pygeos/test/test_constructive.py]: Using numpy `assert_array_equal` here would let you handle 0d and 1d test cases above:

```
assert_array_equal(actual, expected)
```

**Comment by caspervdw** [comment]
Thanks for the review, I updated the PR.
